File: datasets/python_source-code/Multiple-C2/organic_1.py
"""
0/1 Knapsack (DP) - Organic Submission #1
Classic bottom-up 2D DP table.
"""


def knapsack(weights, values, capacity):
    n = len(weights)
    table = [[0] * (capacity + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        for w in range(capacity + 1):
            if weights[i - 1] <= w:
                table[i][w] = max(
                    table[i - 1][w],
                    values[i - 1] + table[i - 1][w - weights[i - 1]],
                )
            else:
                table[i][w] = table[i - 1][w]

    return table[n][capacity]


if __name__ == "__main__":
    weights = [2, 3, 4, 5]
    values = [3, 4, 5, 6]
    capacity = 5
    print("Max value:", knapsack(weights, values, capacity))

File: server/datasets/java_source-code/Multiple-C2/organic_1.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #1
 * Classic bottom-up 2D DP table.
 */

class Solution {
    public static int knapsack(int[] weights, int[] values, int capacity) {
        int n = weights.length;
        int[][] table = new int[n + 1][capacity + 1];

        for (int i = 1; i <= n; i++) {
            for (int w = 0; w <= capacity; w++) {
                if (weights[i - 1] <= w) {
                    table[i][w] = Math.max(
                        table[i - 1][w],
                        values[i - 1] + table[i - 1][w - weights[i - 1]]
                    );
                } else {
                    table[i][w] = table[i - 1][w];
                }
            }
        }
        return table[n][capacity];
    }

    public static void main(String[] args) {
        int[] weights = {2, 3, 4, 5};
        int[] values = {3, 4, 5, 6};
        int capacity = 5;
        System.out.println("Max value: " + knapsack(weights, values, capacity));
    }
}
